### ISSUE TITLE
Axes3D.plot_surface with meshgrid args stopped working

### ISSUE THREAD

**Issue by keszybz** [matplotlib/matplotlib#7313]
To help us understand and resolve your issue please check that you have provided
the information below.

- [x] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...)
Linux, python3-matplotlib-2.0.0-0.2.b4.fc26.x86_64, python3-3.5.2-5.fc26.x86_64

- [x] How did you install Matplotlib and Python (pip, anaconda, from source ...)
Fedora packages (rawhide version).

- [x] If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/)
      that demonstrates the issue i.e a small piece of code which reproduces the issue
      and can be run with out any other (or as few as possible) external dependencies.

```python
# ackley.py
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt

import numpy as np

fig = plt.figure()
ax = Axes3D(fig)
X = np.arange(-30, 30, 0.5)
Y = np.arange(-30, 30, 0.5)
X, Y = np.meshgrid(X, Y)
Z = np.zeros((X.shape[0], Y.shape[0]))

ax.plot_surface(X, Y, Z, rstride=1, cstride=1)

plt.show()
```

Under matplotlib-1.5.2 this shows a plot. Under matplotlib-2 it crashes with:
```pytb
Exception in Tkinter callback
Traceback (most recent call last):
  File "/usr/lib64/python3.5/tkinter/__init__.py", line 1550, in __call__
    return self.func(*args)
  File "/usr/lib64/python3.5/site-packages/matplotlib/backends/backend_tkagg.py", line 283, in resize
    self.show()
  File "/usr/lib64/python3.5/site-packages/matplotlib/backends/backend_tkagg.py", line 354, in draw
    FigureCanvasAgg.draw(self)
  File "/usr/lib64/python3.5/site-packages/matplotlib/backends/backend_agg.py", line 463, in draw
    self.figure.draw(self.renderer)
  File "/usr/lib64/python3.5/site-packages/matplotlib/artist.py", line 63, in draw_wrapper
    draw(artist, renderer, *args, **kwargs)
  File "/usr/lib64/python3.5/site-packages/matplotlib/figure.py", line 1129, in draw
    renderer, self, dsu, self.suppressComposite)
  File "/usr/lib64/python3.5/site-packages/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File "/usr/lib64/python3.5/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 272, in draw
    for col in self.collections]
  File "/usr/lib64/python3.5/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 272, in <listcomp>
    for col in self.collections]
  File "/usr/lib64/python3.5/site-packages/mpl_toolkits/mplot3d/art3d.py", line 626, in do_3d_projection
    cedge = cedge.repeat(len(xyzlist), axis=0)
  File "/usr/lib64/python3.5/site-packages/numpy/ma/core.py", line 2549, in wrapped_method
    result = getattr(self._data, funcname)(*args, **params)
MemoryError
```
It tries to allocate cedge with size 56644 × 14161.

- [ ] If this is an image generation bug attach a screenshot demonstrating the issue.
- [x] If this is a regression (Used to work in an earlier version of Matplotlib), please 
      note where it used to work.

I'd guess that this is a matplotlib-2 regression. python3-matplotlib-1.5.2-0.1.rc2.fc24.x86_64 works fine.

**Comment by tacaswell** [comment]
interesting @WeatherGod do you recall any changes to internal copies?

I can not reproduce this with '2.0.0b4.post208+g6db19a6' but I have a lot of RAM in the comuter.

**Comment by keszybz** [comment]
It crashes more reliably on i686 / armv7hl, but not on amd64 and other 64 bit arches.

**Comment by dopplershift** [comment]
So I've dug in and I think the problem is in this piece of code:
```python
        if len(cface) != len(xyzlist):
            cface = cface.repeat(len(xyzlist), axis=0)
        if len(cedge) != len(xyzlist):
            if len(cedge) == 0:
                cedge = cface
            cedge = cedge.repeat(len(xyzlist), axis=0)
```
This block is unchanged between 1.5.2 and 2.x--however, the default changed. On 1.5.2, with that script, `cedge` is list like `[[1.0, 0.0, 0.0, 0.5]]` (or something). When that goes through that block, it ends up repeated to match the size of `xyzlist` (14161 items).

On 2.x, instead `cedge` is an empty list (due to defaults change taking off edges?). This exposes what I think is a bug in the code, where first `cedge` is set to `cface`, and then is replicated `len(xyzlist)` times. This is unnecessary since cface has already been replicated as necessary. I think the `repeat` on `cedge` should be an `else` clause on `if len(cedge) == 0`.

Can @WeatherGod or someone else double check my logic? Putting the else there allows the script to run, but I'm not sure of the ramifications. The wire and surface demos running locally with that change seem to match those in the dev docs.

**Comment by dopplershift** [comment]
@keszybz Can you try passing `edgecolor='k'` to `plot_surface` and see if that both matches what you expect to see and fixes your problem?

**Comment by keszybz** [comment]
Yep, it works fine with `edgecolor='k'`.

**Comment by WeatherGod** [comment]
No change that I am aware of for 1.5.2 vs. 2.x. Is the same version of
numpy being used in both tests? Also note that there is a logic bug in the
example, but it doesn't exhibit itself because the data is square. The
shape of Z should be the same as the shape of X or Y, not `(X.shape[0],
Y.shape[0])`.

**Comment by keszybz** [comment]
python3-numpy-1.11.0-4.fc24.x86_64
python3-numpy-1.11.2-1.fc26.x86_64
→ so not the same, but similar.

>  Also note that there is a logic bug in the example

This was pared down from a test case in python-deap, it's not real code.

**Comment by dopplershift** [comment]
@WeatherGod I think I have it. Can you double check me on the logic above and the fix in #7348 ?

**Comment by dopplershift** [comment]
Fixed in #7348 .

**Comment by keszybz** [comment]
On Tue, Oct 25, 2016 at 01:09:15PM -0700, Ryan May wrote:
> Fixed in #7348 .

Thanks!
